Attempt to resolve issue 251

eriksiegel opened the pull request:
This turned out to be quite a change due to our decisions that base URIs must be absolute. I had to change `p:archive`, `p:archive-manifest` and `p:unarchive`.

All three now have a `relative-to` option that can be used to set the base URI for the files *in* the archive.

I also made the choice to have all relevant URIs to be normalized first by `p:urify()`. This applies to values passed to the `relative-to` option and the values of the manifest's `c:entry/@href` attributes. IMHO this is in line with the decisions we made about this for setting base URIs on our last call.

eriksiegel commented:
>In https://github.com/xproc/3.0-specification/pull/896/files, I already used D0074ff.

Well, since a week I'm using the error list https://github.com/xproc/3.0-specification/wiki/Error-Codes to get a number for a new error. And this number seemed to be unused... So I suppose you forgot to mention it there?

>Don’t the steps throw C00… errors anyway?

The steps now already specify raising D numbers when it is a common error. And specifying a non absolute base URI seemed a common error to me given the decisions made on last call. Both to be raised by the core *and* the steps. 

>And if p:urify() itself already throws errors if the URI is not absolute and compliant with RFC 3986, do the individual steps have to throw errors?

I didn't realize `p:urify()` raised errors, so you're probably right about steps not having to do that themselves. I'll revisit my changes and see how this affects the descriptions.

gimsieke commented:
> 
> 
> > In https://github.com/xproc/3.0-specification/pull/896/files, I already used D0074ff.
> 
> Well, since a week I'm using the error list https://github.com/xproc/3.0-specification/wiki/Error-Codes to get a number for a new error. And this number seemed to be unused... So I suppose you forgot to mention it there?

I was aware that this is a race condition that may lead to conflicts, but I preferred not to add error messages to the list before the corresponding pull requests have been accepted.

> 
> > Don’t the steps throw C00… errors anyway?
> 
> The steps now already specify raising D numbers when it is a common error. And specifying a non absolute base URI seemed a common error to me given the decisions made on last call. Both to be raised by the core _and_ the steps.

If you say so. I don’t know the system for error prefixes tbh. We need a metaspec for the prefix/numbering scheme ;)

> 
> > And if p:urify() itself already throws errors if the URI is not absolute and compliant with RFC 3986, do the individual steps have to throw errors?
> 
> I didn't realize `p:urify()` raised errors, so you're probably right about steps not having to do that themselves. I'll revisit my changes and see how this affects the descriptions.

Maybe first settle what `p:urify()` does and which errors it throws.

eriksiegel commented:
@gimsieke : I did the following:

* I removed the specific error codes in `p:archive`, `p:archive-manifest` and `p:unarchive` for things that are processed by `p:urify()`. 
* I removed D0074 from the error code wiki page
* There seem to be more/other situations that need error codes, I opened an issue for this to discuss: #256 
* I will put the error code maintenance/policy/claiming/... on the proposed Cologne agenda

My suggestion is to accept the PR with these changes and re-adjust if we find out we do need to say more about specific errors.

gimsieke commented:
I agree that we should discuss it. 

A general rule may be to run some normalization and/or compliance checking function on all `xs:anyURI` options and on the base URIs of all documents.

I’d be hesitant to stipulate that we call `p:urify()` by default in these circumstances. One reason is that it will not be obvious to the processor whether it should be better called with no 2nd argument (which means that relative paths will be resolved against the current working directory if available) or with the context document’s base URI as 2nd argument, or with something else as 2nd argument. This is probably the pipeline author’s call to make.

We should nevertheless stipulate that the processor may (or must?) check that every `xs:anyURI` option and every `base-uri` document property is an absolute URI that complies with RFC 3986.

For `@xml:base`, there is already a rule that it must comply with RCF 3986 (by and large), but it need not be absolute. Since `@xml:base` can be set within an opaque step, for example in `p:xslt`, the processor can’t be bothered IMHO to always check `@xml:base` for compliance. Maybe only when it is set by `p:add-attribute`. I think we can safely defer a check until someone uses this URI for a document property or for an option value.

eriksiegel commented:
You can find the buld step spec at: https://eriksiegel.github.io/3.0-steps/issue-251/head/steps/index.html

Look at the `p:archive`, 'p:archive-manifest' and `p:unarchive`

I included the changes for #256 & #203
